### PR TITLE
[ re #184 ] Axiom.UIP

### PR DESCRIPTION
**Pull request by gallais**

### CODE REVIEW

**Comment by gallais** [comment]
It wasn't as bad as I expected: `Relation.Binary.PropositionalEquality` re-exports the
content of `Core`, so it's practically seamless to transfer things to `Core` to break
dependency cycles!

**Comment by HuStmpHrrr** [comment]
hi, this is very strange: why propositional equality without K should import UIP, despite that `Axiom.UIP` is compiled without K?

**Comment by gallais** [comment]
`Axiom.UIP` does not postulate the axiom, it merely gives its type. It also proves
that for some types, uniqueness of identity proofs holds. It's this second part
which is used in `Relation.Binary.PropositionalEquality`.

**Comment by HuStmpHrrr** [comment]
my particular surprise when seeing this is caused by UIP and that it's treated as an axiom.

I guess I will state my opinion in two ways:

1. UIP/K is a quite special axiom which is admitted by the system, and it does not break computational interpretation like other axioms. so UIP is more like a property in this case, instead of an axiom, and that makes me feel that UIP should be a part of `Relational.Binary.PropositionalEquality` instead of having its own very high level module.

2. a module called `Axiom` is quite scary. I suppose it may start to encourage other axioms to be placed in this module, and importing from `Axiom` should indicate a warning message, indicating that at the very least, computational interpretation no longer works. It's OK if we have `Axiom` modules which contains a number of compatible axioms. However, I think UIP is then in a different situation as explained in the first point, and if we have this style, then, for example, `Decidable` may just become LEM and be put in `Axiom`.

**Comment by gallais** [comment]
* UIP is *not* admitted by the system when `--without-K` is turned on
* `Axiom.UIP` is declared `--safe` and thus it's not possible to add new postulates to it

**Comment by HuStmpHrrr** [comment]
>     * UIP is _not_ admitted by the system when `--without-K` is turned on

This further indicates that UIP should be a property, not an axiom. I am just wondering what led to the choice to have an `Axiom` module and have that to contain UIP.

**Comment by gallais** [comment]
The name originated from the discussion at #184

**Comment by HuStmpHrrr** [comment]
Hmm...

I am uncertain about this. I can only say that importing from `Axiom` will catch eyeballs and raise questions.

**Comment by HuStmpHrrr** [comment]
I guess one idea is to make it `Laws.UIP` to indicate that the module contains consequences of `UIP`. The same applies to other future axiom modules that are safe.

From the discussion, I guess the conclusion is to make the module parameterized and rely on users to postulate the fact. If that's the case, changing the module name to `Laws` is more accurate and more morally correct.

File: CHANGELOG.md [hunk 1]
```diff
@@ -403,6 +406,16 @@ Deprecated features
...
 
+* From `Relation.Binary.PropositionalEquality` to `Relation.Binary.PropositionalEquality.Core`:
+  ```agda
+  cong
+  module ≡-Reasoning
+  trans-reflʳ
+  trans-assoc
+  trans-symˡ
+  trans-symʳ
+  ```
```
Comment (MatthewDaggitt) [inline CHANGELOG.md]: Do we need to mention this in the CHANGELOG? Under the contract that the `Core` modules should never be imported then these changes shouldn't matter?

**Comment by MatthewDaggitt** [comment]
I agree with @gallais that `Axiom` is probably a better name than `Laws`. It should be much clearer once we introduce modules for excluded middle/function extensionality etc.

File: CHANGELOG.md (hunk 1 above)
Comment (gallais) [reply CHANGELOG.md]: Fair enough.

**Pull request merged**